src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -746,18 +753,18 @@ public void close(String reason, boolean flushEngine) throws IOException {
                 }
                 changeState(IndexShardState.CLOSED, reason);
             } finally {
-                final Engine engine = this.currentEngineReference.getAndSet(null);
```

Comment:
Imho I think it's cleaner and better to wait on in flight operations before shutting down. For the same reason we wait on ongoing in engine operations before closing it. I never liked those logged EngineClosed exceptions, even if those are in trace. just my 2cents.